File: python/prims_algorithm.py
import heapq

def prims_mst(graph):
    V = list(graph.keys())
    if not V:
        return []

    start = V[0]
    visited = {start}
    mst = []
    
    # Priority queue stores (weight, u, v)
    edges = [(w, start, v) for v, w in graph.get(start, [])]
    heapq.heapify(edges)

    while edges and len(mst) < len(V) - 1:
        w, u, v = heapq.heappop(edges)

        if v not in visited:
            visited.add(v)
            mst.append((u, v, w))

            for neighbor, weight in graph.get(v, []):
                if neighbor not in visited:
                    heapq.heappush(edges, (weight, v, neighbor))

    return mst

# Example
# Graph represented as an adjacency list: {node: [(neighbor, weight), ...]}
G = {
    'A': [('B', 4), ('H', 8)],
    'B': [('A', 4), ('C', 8), ('H', 11)],
    'C': [('B', 8), ('D', 7), ('F', 4)],
    'D': [('C', 7), ('E', 9), ('F', 14)],
    'E': [('D', 9), ('F', 10)],
    'F': [('C', 4), ('D', 14), ('E', 10), ('G', 2)],
    'G': [('F', 2), ('H', 1), ('I', 6)],
    'H': [('A', 8), ('B', 11), ('G', 1), ('I', 7)],
    'I': [('G', 6), ('H', 7)]
}

mst_edges = prims_mst(G)
print(f"Prim's MST Edges: {mst_edges}")

File: go/prims_algorithm.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

// Edge represents an edge to a neighbor with a weight
type Edge struct {
	to, weight int
}

// PriorityQueueItem for min-heap
type PriorityQueueItem struct {
	node, weight int
}

// Min-heap implementation
type PriorityQueue []PriorityQueueItem

func (pq PriorityQueue) Len() int { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].weight < pq[j].weight
}
func (pq PriorityQueue) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i] }

func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(PriorityQueueItem))
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

// Prim's Algorithm
func primMST(graph [][]Edge, start int) int {
	n := len(graph)
	visited := make([]bool, n)
	pq := &PriorityQueue{}
	heap.Init(pq)

	heap.Push(pq, PriorityQueueItem{node: start, weight: 0})
	totalWeight := 0

	for pq.Len() > 0 {
		cur := heap.Pop(pq).(PriorityQueueItem)
		u := cur.node
		w := cur.weight

		if visited[u] {
			continue
		}

		visited[u] = true
		totalWeight += w

		for _, edge := range graph[u] {
			if !visited[edge.to] {
				heap.Push(pq, PriorityQueueItem{node: edge.to, weight: edge.weight})
			}
		}
	}

	return totalWeight
}

func main() {
	/*
		Graph (undirected weighted):
		0 --2-- 1
		0 --3-- 3
		1 --2-- 2
		1 --4-- 3
		2 --1-- 3
	*/

	graph := [][]Edge{
		{{1, 2}, {3, 3}},    // 0
		{{0, 2}, {2, 2}, {3, 4}}, // 1
		{{1, 2}, {3, 1}},    // 2
		{{0, 3}, {1, 4}, {2, 1}}, // 3
	}

	total := primMST(graph, 0)
	fmt.Println("Total weight of MST:", total)
}
